core/cthreads.c

```diff
@@ -0,0 +1,291 @@
+    pthread_mutex_t pMutex = PTHREAD_MUTEX_INITIALIZER;
+    pthread_mutex_init(&pMutex, NULL);
+
+    mutex.pMutex = pMutex;
+  #elif _WIN32
+    HANDLE wMutex = CreateMutex(NULL, FALSE, NULL);
+
+    mutex.wMutex = wMutex;
+  #endif
+
+  return mutex;
+}
+
+int cthreads_mutex_init(struct cthreads_mutex *mutex) {
+  #ifdef __unix__
+    pthread_mutex_t pMutex = PTHREAD_MUTEX_INITIALIZER;
+    int res = pthread_mutex_init(&pMutex, NULL);
+
+    mutex->pMutex = pMutex;
+
+    return res;
+  #elif _WIN32
+    HANDLE wMutex = CreateMutex(NULL, FALSE, NULL);
+
+    if (wMutex == NULL) return 1;
+    else {
+      mutex->wMutex = wMutex;
+      return 0;
+    }
+  #endif
+}
+
+int cthreads_mutex_lock(struct cthreads_mutex *mutex) {
+  #ifdef __unix__
+    return pthread_mutex_lock(&mutex->pMutex);
+  #elif _WIN32
+    DWORD ret = WaitForSingleObject(mutex->wMutex, INFINITE);
+
+    if (ret == WAIT_OBJECT_0) return 0;
+    else return 1;
+  #endif
+}
+
+int cthreads_mutex_trylock(struct cthreads_mutex *mutex) {
+  #ifdef __unix__
+    return pthread_mutex_trylock(&mutex->pMutex);
+  #elif _WIN32
+    return WaitForSingleObject(mutex->wMutex, 0);
+  #endif
+}
+
+int cthreads_mutex_unlock(struct cthreads_mutex *mutex) {
+  #ifdef __unix__
+    return pthread_mutex_unlock(&mutex->pMutex);
+  #elif _WIN32
+    ReleaseMutex(mutex->wMutex);
+
+    return 0;
+  #endif
+}
+
+void cthreads_mutex_destroy(struct cthreads_mutex *mutex) {
+  #ifdef __unix__
+    pthread_mutex_destroy(&mutex->pMutex);
+  #elif _WIN32
+    CloseHandle(mutex->wMutex);
+  #endif
+}
+
+int cthreads_cond_init(struct cthreads_cond *cond) {
+  #ifdef __unix__
+    pthread_cond_t pCond;
+    int res = pthread_cond_init(&pCond, NULL);
+
+    cond->pCond = pCond;
+
+    return res;
+  #elif _WIN32
+    HANDLE wCond = CreateEvent(NULL, FALSE, FALSE, NULL);
+
+    cond->wCond = wCond;
+
+    return 0;
+  #endif
+}
+
+int cthreads_cond_signal(struct cthreads_cond *cond) {
+  #ifdef __unix__
+    return pthread_cond_signal(&cond->pCond);
+  #elif _WIN32
+    SetEvent(cond->wCond);
+
+    return 0;
+  #endif
+}
+
+int cthreads_cond_broadcast(struct cthreads_cond *cond) {
+  #ifdef __unix__
+    return pthread_cond_broadcast(&cond->pCond);
+  #elif _WIN32
+    SetEvent(cond->wCond);
+
+    return 0;
+  #endif
+}
+
+int cthreads_cond_destroy(struct cthreads_cond *cond) {
+  #ifdef __unix__
+    return pthread_cond_destroy(&cond->pCond);
+  #elif _WIN32
+    CloseHandle(cond->wCond);
+
+    return 0;
+  #endif
+}
+
+int cthreads_cond_wait(struct cthreads_cond *cond, struct cthreads_mutex *mutex) {
+  #ifdef __unix__
+    return pthread_cond_wait(&cond->pCond, &mutex->pMutex);
+  #elif _WIN32
+    cthreads_mutex_unlock(mutex);
+    WaitForSingleObject(cond->wCond, INFINITE);
+    cthreads_mutex_lock(mutex);
+
+    return 0;
+  #endif
+}
+
+int cthreads_wait_close_thread(struct cthreads_thread *thread, void *code) {
+  #ifdef __unix__
+    return pthread_join(thread->pThread, &code);
```

Comment:
I don't think this is going to work in the case where we need a `NULL` as the last parameter... Currently, it is impossible to pass a NULL value, so maybe this?
```suggestion
    return pthread_join(thread->pThread, code ? &code : NULL);
```